full-moon/src/ast/parsers.rs

```diff
@@ -1663,24 +1926,935 @@ fn parse_function_body(state: &mut ParserState) -> ParserResult<FunctionBody> {
         parameters.push(Pair::End(last_parameter.into_value()));
     }
 
+    #[cfg(feature = "roblox")]
+    let return_type = if state.lua_version().has_luau() {
+        // rewrite todo: we should emit a warning if the user tried to use `->` instead of `:`
+        if let Some(punctuation) = state.consume_if(Symbol::Colon) {
+            match parse_return_type(state) {
+                ParserResult::Value(type_info) => Some(ast::TypeSpecifier {
+                    punctuation,
+                    type_info,
+                }),
+                _ => return ParserResult::LexerMoved,
+            }
+        } else {
+            None
+        }
+    } else {
+        None
+    };
+
     let (block, end) = match parse_block_with_end(state, "function body", &right_parenthesis) {
         Ok((block, end)) => (block, end),
         Err(()) => return ParserResult::LexerMoved,
     };
 
     ParserResult::Value(FunctionBody {
+        #[cfg(feature = "roblox")]
+        generics,
         parameters_parentheses: ContainedSpan::new(left_parenthesis, right_parenthesis),
         parameters,
+        #[cfg(feature = "roblox")]
+        type_specifiers,
+        #[cfg(feature = "roblox")]
+        return_type,
         block,
         end_token: end,
     })
 }
 
+#[cfg(feature = "luau")]
+fn expect_if_else_expression(
+    state: &mut ParserState,
+    if_token: TokenReference,
+) -> Result<ast::IfExpression, ()> {
+    let ParserResult::Value(condition) = parse_expression(state) else {
+        return Err(())
+    };
+
+    let Some(then_token) = state.require(Symbol::Then, "expected `then` after condition") else {
+        return Err(())
+    };
+
+    let ParserResult::Value(if_expression) = parse_expression(state) else {
+        return Err(())
+    };
+
+    let mut else_if_expressions = Vec::new();
+    while let Some(else_if_token) = state.consume_if(Symbol::ElseIf) {
+        let ParserResult::Value(condition) = parse_expression(state) else {
+            return Err(())
```

Comment:
Hmm, we could maybe recover these but then it'd just error when it eventually hits the `else` anyway

Reply:
Yeah i'm not too sure about this. If there's an example of something that would be nice if recoverable then I can take a look

Reply:
Actually, https://github.com/Kampfkarren/full-moon/issues/265 is a similar example (except at the else block) where this is present.

I think we need some way to add "error terminals" to the graph. Like AstExprError or AstTypeError